packages/cosmic-swingset/src/anylogger-agd.js

```diff
@@ -0,0 +1,165 @@
+};
+
+/**
+ * @type {Record<string,
+ *   (opts: { now: Date, label: string, level: LogLevel, args: any[] }) => any[]
+ * >}
+ */
+const logFormatterForLogFormat = {
+  legacy: ({ now, label, args }) => {
+    const timestamp = now.toISOString();
+    return [`${timestamp} ${label}:`, ...args];
+  },
+  plain: ({ now, label, level, args }) => {
+    const timestamp = plainTime(now);
+    const plainLevel = levelToPlainLevel[level];
+    return [`${timestamp} ${plainLevel} ${label}:`, ...args];
+  },
+  json: ({ now, label, level, args }) => {
+    const message = [`${label}:`, ...args]
+      .map(a => (typeof a === 'string' ? a : util.inspect(a)))
+      .join(' ');
+    return [
+      JSON.stringify({
+        level,
+        time: dateToNumericOffsetString(now),
+        message,
+      }),
+    ];
+  },
+};
+
+/** @type {string} */
+let logFormat;
+/** @type {typeof logFormatterForLogFormat[string]} */
+let logFormatter;
+
+/** @param {string} format */
+export const setLogFormat = format => {
+  logFormat = format;
+  logFormatter =
+    logFormatterForLogFormat[logFormat] || logFormatterForLogFormat.plain;
+};
+setLogFormat('plain'); // default
+
+//#endregion
+
+// Turn on debugging output with DEBUG=agoric or DEBUG=agoric:${level}
+const DEBUG_LIST = getEnvironmentOptionsList('DEBUG');
+
+/**
+ * As documented in ../../../docs/env.md, the log level defaults to "log" when
+ * environment variable DEBUG is non-empty or unset, and to the quieter "info"
+ * when it is set to an empty string, but in either case is overridden if DEBUG
+ * is a comma-separated list that contains "agoric:none" or "agoric:${level}" or
+ * "agoric" (the last an alias for "agoric:debug").
+ *
+ * @type {string | undefined}
+ */
+let maxActiveLevel =
+  DEBUG_LIST.length > 0 || getEnvironmentOption('DEBUG', 'unset') === 'unset'
+    ? 'log'
+    : 'info';
+for (const selector of DEBUG_LIST) {
+  const fullSelector = selector === 'agoric' ? 'agoric:debug' : selector;
+  const [_, detail] = fullSelector.match(/^agoric:(.*)/gs) || [];
+  if (detail) {
+    maxActiveLevel = detail === 'none' ? undefined : detail;
+  }
+}
+const maxActiveLevelCode = /** @type {number} */ (
+  (maxActiveLevel && anylogger.levels[maxActiveLevel]) ?? -Infinity
+);
+
+const oldExt = anylogger.ext;
+anylogger.ext = logger => {
+  logger = oldExt(logger);
+
+  /** @type {(level: LogLevel) => boolean} */
+  const enabledFor = level => anylogger.levels[level] <= maxActiveLevelCode;
+  logger.enabledFor = enabledFor;
+
+  const nameColon = `${logger.name}:`;
+  const label = logger.name.replaceAll(':', ': ');
+
+  // Vat logs are suppressed unless matched by a prefix in DEBUG_LIST.
+  const suppressed =
+    VAT_LOGGER_PREFIXES.some(prefix => nameColon.startsWith(`${prefix}:`)) &&
+    !DEBUG_LIST.some(prefix => nameColon.startsWith(`${prefix}:`));
+
+  const levels = /** @type {LogLevel[]} */ (Object.keys(anylogger.levels));
+  for (const level of levels) {
```

Comment:
This is pretty tricky stuff. I don't see unit tests for it. I wouldn't be comfortable maintaining it without tests.